boston-crime-data/content.md

```diff
@@ -0,0 +1,379 @@
+```python
+scaler = StandardScaler()
+scaler.fit(data)
+y_mean = scaler.mean_[target_pos]
+y_var = scaler.var_[target_pos]
+y_std = sqrt(y_var)
+data = scaler.transform(data)
+```
+
+We split the data into training, validation and test sets
+```python
+x_train = data[:train_size, :]
+x_train = np.reshape(x_train, shape)
+y_train = x_train[:, 1:, target_pos:target_pos + 1]
+x_train = x_train[:, :-1]
+
+x_val = data[train_size:train_size + val_size, :]
+x_val = np.reshape(x_val, shape)
+y_val = x_val[:, 1:, target_pos:target_pos + 1]
+x_val = x_val[:, :-1]
+
+x_test = data[train_size + val_size:train_size + val_size + test_size, :]
+x_test = np.reshape(x_test, shape)
+y_test = x_test[:, 1:, target_pos:target_pos + 1]
+x_test = x_test[:, :-1]
+```
+
+Now we loop over a few hyperparameters to see which configuration gives the best result. It is interesting to play around with
+nodes, dropout and l_rate, also i, j, k and epochs can be adjusted for quicker exploratory runs. Still RNNs run slowly but can
+be sped up with gpu computing.
+```python
+result_list = []
+for i in range(2):
+    for j in range(2):
+        for k in range(2):
+            # set hyperparameters
+            nodes = 130 + 20*i
+            dropout = 0.2 * j
+            l_rate = 0.001 + 0.005 * k
+            # specify optimizer in order to control the learning rate
+            optimizer = Adam(lr=l_rate)
+            # build a shallow RNN
+            model1 = Sequential()
+            model1.add(LSTM(nodes, input_shape=(None, input_width), return_sequences=True, dropout=dropout, recurrent_dropout=0))
+            model1.add(Dense(output_width))
+            model1.compile(
+                    optimizer=optimizer, loss='mse')
+            # use early stopping and a checkpoint to pick the best model before overfitting starts happening
+            stop = EarlyStopping(patience=20)
+            checkpoint = ModelCheckpoint('best_model.h5', save_best_only=True)
+
+            # fit the model
+            history1 = model1.fit(x_train, y_train, epochs=200, validation_data=(x_val, y_val), callbacks=[stop, checkpoint])
+
+            # print some information about each model
+            network_prediction = model1.predict(x_test)
+            network_prediction = network_prediction.reshape((-1)) * y_std + y_mean
+            # invert the standardization before measuring error
+            y_test_post = y_test * y_std + y_mean
+            y_test_post = y_test_post.reshape(-1)
+            network_score = MSE(network_prediction, y_test_post)
+            print('Our prediction has MSE: ', network_score)
+            print('with dropout ', dropout)
+            print('hidden layer nodes ', nodes)
+            print('and learning rate ', l_rate)
+            result_list.append([network_score, dropout, nodes, l_rate])
+
+            # log model to mflux
+            model1_score = model1.evaluate(x_test, y_test)
+            mflux_ai.init("YOUR CODE HERE")
```

Comment:
Use "your_project_token_goes_here" instead. mflux_ai checks if the token contains "your_" and will help the user if it does :)